compiler/src/iree/compiler/Codegen/Common/ForallToFor.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "llvm/ADT/SmallVectorExtras.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-codegen-forall-to-for"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_FORALLTOFORPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+namespace {
+
+LogicalResult forallToForLoop(RewriterBase &rewriter, scf::ForallOp forallOp) {
+  rewriter.setInsertionPoint(forallOp);
+
+  SmallVector<Value> lbs = forallOp.getLowerBound(rewriter);
+  SmallVector<Value> ubs = forallOp.getUpperBound(rewriter);
+  SmallVector<Value> steps = forallOp.getStep(rewriter);
+  SmallVector<Value> iterArgs = forallOp.getOutputs();
+
+  bool unsupportedOperation = false;
+  auto buildBody = [&](OpBuilder &builder, Location loc, ValueRange ivs,
+                       ValueRange args) -> scf::ValueVector {
+    // Inline `scf.forall` body excluding terminating
+    // `scf.forall.in_parallel` op.
+    IRMapping map;
+    map.map(forallOp.getInductionVars(), ivs);
+    map.map(forallOp.getRegionOutArgs(), args);
+    for (auto &op : forallOp.getBody()->without_terminator()) {
+      builder.clone(op, map);
+    }
+
+    // Convert + inline the contents of `scf.forall.in_parallel` terminator.
+    auto terminator = forallOp.getTerminator();
+    SmallVector<Value> yieldedValues;
+    for (auto &yieldOp : terminator.getYieldingOps()) {
+      // Convert tensor.parallel_insert_slice to tensor.insert_slice
+      if (auto parallelInsert =
+              dyn_cast<tensor::ParallelInsertSliceOp>(&yieldOp)) {
+
+        auto insertSlice = builder.create<tensor::InsertSliceOp>(
+            parallelInsert.getLoc(),
+            map.lookupOrDefault(parallelInsert.getSource()),
+            map.lookupOrDefault(parallelInsert.getDest()),
+            llvm::map_to_vector(
+                parallelInsert.getOffsets(),
+                [&](Value v) -> Value { return map.lookupOrDefault(v); }),
+            llvm::map_to_vector(
+                parallelInsert.getSizes(),
+                [&](Value v) -> Value { return map.lookupOrDefault(v); }),
+            llvm::map_to_vector(
+                parallelInsert.getStrides(),
+                [&](Value v) -> Value { return map.lookupOrDefault(v); }),
+            parallelInsert.getStaticOffsets(), parallelInsert.getStaticSizes(),
+            parallelInsert.getStaticStrides());
```

Comment:
```suggestion

auto source = map.lookupOrDefault(parallelInsert.getSource());
auto dest = map.lookupOrDefault(parallelInsert.getDest());
auto offsets = llvm::map_to_vector(parallelInsert.getOffsets(),
    [&](Value v) { return map.lookupOrDefault(v); });
auto sizes = llvm::map_to_vector(parallelInsert.getSizes(),
    [&](Value v) { return map.lookupOrDefault(v); });
auto strides = llvm::map_to_vector(parallelInsert.getStrides(),
    [&](Value v) { return map.lookupOrDefault(v); });
auto insertSlice = builder.create<tensor::InsertSliceOp>(
     parallelInsert.getLoc(), source, dest, offsets, sizes, strides,
    parallelInsert.getStaticOffsets(),
    parallelInsert.getStaticSizes(),
    parallelInsert.getStaticStrides());
```
 I think it's cleaner to have the argument outside the builder and then pass but that's just an opinion.